js/ai/src/prompt.ts

```diff
@@ -137,79 +174,381 @@ export interface ExecutablePrompt<
   asTool(): Promise<ToolAction>;
 }
 
+export type PartsGenerator<I, S = any> = (
+  input: I,
+  options: {
+    state?: S;
+  }
+) => Part[] | Promise<string | Part | Part[]>;
+
+export type MessageGenerator<I, S = any> = (
+  input: I,
+  options: {
+    history?: MessageData[];
+    state?: S;
+  }
+) => MessageData[] | Promise<MessageData[]>;
+
+export type DocumentGenerator<I, S = any> = (
+  input: I,
+  options: {
+    state?: S;
+  }
+) => DocumentData[] | Promise<DocumentData[]>;
+
+interface PromptCache {
+  userPrompt?: PromptFunction;
+  system?: PromptFunction;
+  messages?: PromptFunction;
+}
+
 /**
- * Defines and registers a prompt action. The action can be called to obtain
- * a `GenerateRequest` which can be passed to a model action. The given
- * `PromptFn` can perform any action needed to create the request such as rendering
- * a template or fetching a prompt from a database.
+ * Defines a prompt which can be used to generate content or render a request.
  *
- * @returns The new `PromptAction`.
+ * @returns The new `ExecutablePrompt`.
  */
-export function definePrompt<I extends z.ZodTypeAny>(
+export function definePrompt<
+  I extends z.ZodTypeAny = z.ZodTypeAny,
+  O extends z.ZodTypeAny = z.ZodTypeAny,
+  CustomOptions extends z.ZodTypeAny = z.ZodTypeAny,
+>(
   registry: Registry,
-  config: PromptConfig<I>,
-  fn: PromptFn<I>
-): PromptAction<I> {
-  const a = defineAction(
+  options: PromptConfig<I, O, CustomOptions>
+): ExecutablePrompt<z.infer<I>, O, CustomOptions> {
+  const promptCache = {} as PromptCache;
+
+  const renderOptionsFn = async (
+    input: z.infer<I>,
+    renderOptions: PromptGenerateOptions<O, CustomOptions> | undefined
+  ): Promise<GenerateOptions> => {
+    const messages: MessageData[] = [];
+    renderOptions = { ...renderOptions }; // make a copy, we will be trimming
+    const session = getCurrentSession(registry);
+
+    // order of these matters:
+    await renderSystemPrompt(
+      registry,
+      session,
+      input,
+      messages,
+      options,
+      promptCache
+    );
+    await renderMessages(
+      registry,
+      session,
+      input,
+      messages,
+      options,
+      renderOptions,
+      promptCache
+    );
+    await renderUserPrompt(
+      registry,
+      session,
+      input,
+      messages,
+      options,
+      promptCache
+    );
+
+    let docs: DocumentData[] | undefined;
+    if (typeof options.docs === 'function') {
+      docs = await options.docs(input, {
+        state: session?.state,
+      });
+    } else {
+      docs = options.docs;
+    }
```

Comment:
Not necessarily a now thing, but I feel like we should have a generic "resolver" utility that can do this kind of "if func pass it some stuff and await the response else take the literal value". We use that all over the place.

Reply:
yes, but those resolver functions are not all the same... 2 are (for now) but messages is different.